src/mpi/init/init_async.c

```diff
@@ -107,6 +107,8 @@ int MPIR_Init_async_thread(void)
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
+    /* Asynchronous progress require THREAD_MULTIPLE */
+    MPIR_Assert(MPIR_ThreadInfo.isThreaded);
```

Comment:
I don't think we can do this. It is not the user's responsibility to require `MPI_THREAD_MULTIPLE`  when using async progress.

Reply:
Using async progress is an explicit user option, and whether to use `MPI_THREAD_MULTIPLE` is also a user option. I think it is user's responsibility to ensure required options are consistent. Right now we change the thread-level option to `MPI_THREAD_MULTIPLE` behind the user, without letting user know. What if the user really want `MPI_THREAD_SINGLE`, just didn't realize the async progress won't work in `MPI_THREAD_SINGLE` and he perfers to have `MPI_THREAD_SINGLE` instead of async progress?

Reply:
Async progress might not require MPI_THREAD_MULTIPLE. We cannot require the user to do this. It is wrong.

Reply:
I think it is wrong the other way around. When user request for MPI_THREAD_SINGLE, we give them MPI_THREAD_MULTIPLE, is this allowed in standard? Regardless, as a user, I find it very counter-intuitive.

Reply:
> Async progress might not require MPI_THREAD_MULTIPLE.

How?

Reply:
Device-level async progress. Nevertheless, this is long-standing MPICH behavior. Changing it is the wrong thing to do. Please remove it.

Reply:
With device-level async progress (I assume it exist and works), the entire upper-layer async progress will/should be configured out, which will result in asynch-progress work with `MPI_THREAD_SINGLE`. That is the behavior we intended, right? The code right now will make the thread-level into `MPI_THREAD_MULTIPLE` even with device-level async progress. 

Sure. I can remove the commit.

Reply:
> I think it is wrong the other way around. When user request for MPI_THREAD_SINGLE, we give them MPI_THREAD_MULTIPLE, is this allowed in standard? Regardless, as a user, I find it very counter-intuitive.

Yes, it is allowed in the standard. I'm not sure how it would be possible to disallow...